librz/analysis/arch/8051/8051_op.h

```diff
@@ -0,0 +1,614 @@
+};
+// clang-format on
+
+typedef enum {
+	I8051_ADDRESSING_REGISTER = 0,
+	I8051_ADDRESSING_DIRECT,
+	I8051_ADDRESSING_INDIRECT,
+	I8051_ADDRESSING_IMMEDIATE,
+	I8051_ADDRESSING_IMMEDIATE16,
+	I8051_ADDRESSING_RELATIVE,
+	I8051_ADDRESSING_ABSOLUTE,
+	I8051_ADDRESSING_LONG,
+	I8051_ADDRESSING_INDEXED,
+
+	I8051_ADDRESSING_BIT,
+} I8051AddressingMode;
+
+typedef enum {
+	I8051_R0 = 0,
+	I8051_R1,
+	I8051_R2,
+	I8051_R3,
+	I8051_R4,
+	I8051_R5,
+	I8051_R6,
+	I8051_R7,
+	I8051_SP = 0x81,
+	I8051_DPTR,
+	I8051_PCON = 0x87,
+	I8051_TCON,
+	I8051_TMOD,
+	I8051_TL0,
+	I8051_TL1,
+	I8051_TH0,
+	I8051_TH1,
+	I8051_PC,
+	I8051_PSW = 0xD0,
+	I8051_Z = 0xD1,
+	I8051_OV = 0xD2,
+	I8051_RS0 = 0xD3,
+	I8051_RS1 = 0xD4,
+	I8051_N = 0xD5,
+	I8051_AC = 0xD6,
+	I8051_CY = 0xD7,
+	I8051_A = 0xE0,
+	I8051_B = 0xF0
+} I8051Registers;
+
+typedef struct i8051_op_addressing_t {
+	ut64 pc;
+	I8051AddressingMode mode;
+	union {
+		I8051Registers reg;
+		ut8 addr;
+		ut16 addr16;
+		ut16 constant;
+		struct i8051_op_addressing_t *indirect;
+	} d;
+} I8051OpAddressing;
+
+typedef struct i8051_op_t {
+	ut64 pc;
+	ut8 opcode;
+	ut8 len;
+	I8051OpInst inst;
+	ut8 argc;
+	I8051OpAddressing **argv;
+} I8051Op;
+
+static I8051OpAddressing *addressing_addr(I8051AddressingMode mode, ut16 addr) {
+	I8051OpAddressing *a = RZ_NEW0(I8051OpAddressing);
+	if (!a) {
+		return NULL;
+	}
+	a->mode = mode;
+	a->d.addr16 = addr;
+	return a;
+}
+
+static I8051OpAddressing *addressing_direct(ut8 addr) {
+	return addressing_addr(I8051_ADDRESSING_DIRECT, addr);
+}
+
+static I8051OpAddressing *addressing_indirect(I8051OpAddressing *addr) {
+	I8051OpAddressing *a = RZ_NEW0(I8051OpAddressing);
+	if (!a) {
+		return NULL;
+	}
+	a->mode = I8051_ADDRESSING_INDIRECT;
+	a->d.indirect = addr;
+	return a;
+}
+
+static I8051OpAddressing *addressing_relative(ut8 addr) {
+	return addressing_addr(I8051_ADDRESSING_RELATIVE, addr);
+}
+
+static I8051OpAddressing *addressing_immediate(ut8 imm) {
+	return addressing_addr(I8051_ADDRESSING_IMMEDIATE, imm);
+}
+
+static I8051OpAddressing *addressing_bit(ut8 addr) {
+	return addressing_addr(I8051_ADDRESSING_BIT, addr);
+}
+
+static I8051OpAddressing *addressing_register(I8051Registers reg) {
+	I8051OpAddressing *a = RZ_NEW0(I8051OpAddressing);
+	if (!a) {
+		return NULL;
+	}
+	a->mode = I8051_ADDRESSING_REGISTER;
+	a->d.reg = reg;
+	return a;
+}
+
+static inline I8051OpAddressing *addressing_register_a() {
+	return addressing_register(I8051_A);
+}
+
+static inline I8051OpAddressing *addressing_indexed(I8051Registers reg) {
+	I8051OpAddressing *a = RZ_NEW0(I8051OpAddressing);
+	if (!a) {
+		return NULL;
+	}
+	a->mode = I8051_ADDRESSING_INDEXED;
+	a->d.reg = reg;
+	return a;
+}
+
+static inline I8051OpAddressing *addressing_register_b() {
+	return addressing_register(I8051_B);
+}
+
+/**
+ * 0x05 iram addr
+ * 0x06-0x07 @(R0-R1)
+ * 0x08-0x0f, R0-R7
+ */
```

Comment:
Is the meaning here common knowledge for people who know 8051? Otherwise I would suggest to add more context (this is the offset into the buffer?).

Reply:
Same for comments below.